keel-core/src/main/kotlin/com/netflix/spinnaker/keel/constraints/AllowedTimesConstraintEvaluator.kt

```diff
@@ -0,0 +1,228 @@
+package com.netflix.spinnaker.keel.constraints
+
+import com.netflix.spinnaker.keel.api.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.TimeWindowConstraint
+import com.netflix.spinnaker.keel.constraints.ConstraintEvaluator.Companion.getConstraintForEnvironment
+import com.netflix.spinnaker.keel.exceptions.InvalidConstraintException
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import org.springframework.stereotype.Component
+import java.text.ParsePosition
+import java.time.Clock
+import java.time.DayOfWeek
+import java.time.ZoneId
+import java.time.format.DateTimeFormatter
+import java.time.format.TextStyle
+import java.util.Locale
+
+/**
+ * An environment promotion constraint to gate promotions to time windows.
+ * Multiple windows can be specified.
+ *
+ * Hours and days parameters support a mix of ranges and comma separated values.
+ *
+ * Days can be specified using full or short names according to the default JVM locale.
+ *
+ * A timezone can be set in the constraint and a system-wide default can by set via
+ * the dynamic property "default.time-zone" which defaults to UTC.
+ *
+ * Example env constraint:
+ *
+ * constraints:
+ *  - type: allowed-times
+ *    windows:
+ *      - days: Monday,Tuesday-Friday
+ *        hours: 11-16
+ *        tz: America/Los_Angeles
+ *      - days: Wed
+ *        hours: 12-14
+ *        tz: America/Los_Angeles
+ */
+@Component
```

Comment:
This class is super clear. Thanks for adding great class docs, useful helper functions, keeping the core logic short and easy to reason about, and pulling the complex parsing out into smaller functions. 

I struggled with this kind of calculation in Swabbie, and it's delightful to see something that reads well and makes sense here!